papers/models.py

```diff
@@ -697,6 +697,14 @@ def cache_oairecords(self):
         """
         self.cached_oairecords = list(self.oairecord_set.all())
 
+    @property
+    def sorted_oairecords_with_published_deposit(self):
+        """
+        Fetches all oairecords; if there's a DepositRecord, require that it is published
+        """
+        l = [r for r in self.sorted_oai_records if not r.depositrecord_set.exists() or r.depositrecord_set.filter(status='published')]
```

Comment:
That is likely to generate a lot of SQL requests, no? Perhaps it'd be more efficient to use the visibility field of the OaiRecord model: make sure we mark them as invisible when we create OaiRecords that are not published yet, and update the visibility once the deposit succeeds. That will also require a migration to set the visibility on existing pending deposits.

Reply:
Yes, it does. Alternatively I can design a single query with `prefetch_related` and not using `self.sorted_oai_records`.

Visibility is a field of Paper model, not of OaiRecord model. It was my first intention, too.

Reply:
Ah makes sense. I don't know if it is worth adding a field on OaiRecord for that, then. Perhaps `prefetch_related` is good enough in that case then.

Reply:
Prefetch related does two queries anyway, so we can still uses cached sorted oairecords and fetch the DepositRecords for them after.